query/src/main/scala/filodb/query/exec/MetadataExecPlan.scala

```diff
@@ -395,75 +419,214 @@ final case class LabelCardinalityExec(queryContext: QueryContext,
+  case class CardCounts(active: Int, total: Int) {
+    require(total >= active, "total must be at least as large as active")
+  }
 }
 
 /**
-  * Given a shardKeyPrefix, retrieves an estimate of the top-k label-values
-  *  from a specific shard. This is an "estimate" because sketches are used to
-  *  contain count data.
-  *
-  * The accuracy of the estimate is tunable by MAX_ITEMSKETCH_MAP_SIZE.
-  *   The link at its definition describes how its value will affect the estimate.
-  *
-  * Note that the sketch-- a DataSketches ItemsSketch-- maintains a set of
-  *   "heavy hitters." These are the elements whose counts lie above
-  *   a threshold that increases as any counts are increased. This means the
-  *   data must be "sufficiently skewed" in order for a sketch to show that
-  *   any of its elements were heavy hitters.
-  *
-  * Additionally, MAX_ITEMSKETCH_MAP_SIZE bounds the number of elements the
-  *   sketch counts at any given time. If sketch.update(elt, count) is called
-  *   on a currently-uncounted element while the sketch is at full capacity, then
-  *   all counters are decreased by some median count-value, and any below zero
-  *   are removed to make room for the new element. Again: tread carefully when
-  *   metric cardinalities are similar.
-  *
-  * See here for more details:
-  *   https://datasketches.apache.org/docs/Frequency/FrequentItemsOverview.html
+  * Creates a map of (prefix -> cardinalities) according to the TsCardinalities LogicalPlan.
+  *   See TsCardinalities for more details.
   */
-final case class TopkCardExec(queryContext: QueryContext,
-                              dispatcher: PlanDispatcher,
-                              dataset: DatasetRef,
-                              shard: Int,
-                              shardKeyPrefix: Seq[String],
-                              k: Int,
-                              addInactive: Boolean) extends LeafExecPlan {
+final case class TsCardExec(queryContext: QueryContext,
+                            dispatcher: PlanDispatcher,
+                            dataset: DatasetRef,
+                            shard: Int,
+                            shardKeyPrefix: Seq[String],
+                            groupDepth: GroupDepth.Value) extends LeafExecPlan {
+  require(1 + groupDepth.prefixIndex >= shardKeyPrefix.size,
+    "groupDepth indicate a depth at least as deep as shardKeyPrefix")
 
   override def enforceLimit: Boolean = false
 
+  /**
+   * Given a shard key prefix, iterates through all child
+   *   prefixes with the specified size.
+   *
+   * Note: next() will always return the same array instance.
+   *
+   * @param tsMemStore contains the shard keys over which to iterate
+   * @param initPrefix shard key prefix to resolve
+   * @param extendToSize the length of the prefix children to iterate through.
+   *   Must be at (1) least initPrefix.size and (2) greater than zero.
+   */
+  class PrefixIterator (tsMemStore: TimeSeriesMemStore,
```

Comment:
This is repurposed from an older piece of code. Its original intent was to keep as little in memory as possible while expanding shard key prefixes. It will do that same job here, but its benefit is probably less substantial since every prefix needs to be kept in-memory, now, anyways. Let me know if I should trash this in favor of something less complicated.